### ISSUE TITLE
Speed up stable sorting

### ISSUE THREAD

**Issue by oisdk** [haskell/containers#491]
Since the unstable sort is significantly quicker than the stable variant, I think you can write a stable sort that uses it internally, and the overhead of "stabilising" it should be small enough that it's still faster than Data.List.sortBy.

I've [written an implementation](https://github.com/oisdk/containers/tree/faster-stable-sort) which uses a pairing heap which tags each element with its original position:

```haskell
data PQS e = PQS {-# UNPACK #-} !Int e (PQSL e)
data PQSL e = Nl | {-# UNPACK #-} !(PQS e) :&& PQSL e
```

It's quicker than the old implementation (which used Data.List.sort) in the general case (all times are on sequences of length 50000):

```
Times (ms)            min    est    max  std dev   r²
to/from list:        64.23  64.50  64.81  0.432  1.000
1/11/18 stable heap: 38.87  39.40  40.09  0.457  0.999
```

Slightly slower in the case of already sorted lists:

```
Times (ms)            min    est    max  std dev   r²
to/from list:        6.806  6.861  6.901  0.234  1.000
1/11/18 stable heap: 8.211  8.268  8.328  0.111  1.000
```

And quicker in the case of lists sorted in reverse:

```
Times (ms)            min    est    max  std dev   r²
to/from list:        26.79  28.34  30.55  1.219  0.988
1/11/18 stable heap: 9.405  10.13  10.91  0.670  0.977
```

Interestingly, the stable sort is now competitive with the unstable:

```
Times (ms)            min    est    max  std dev   r²
unstable:            34.71  35.10  35.38  0.845  1.000
stable:              38.84  39.22  39.59  0.564  0.999
```

And even beats it in the case of already-sorted lists:

```
Times (ms)            min    est    max  std dev   r²
unstable:            8.457  8.499  8.536  0.069  1.000
stable:              8.160  8.230  8.334  0.158  0.999
```

I couldn't think of an way to actually use `unstableSort`: the new stable sort tags elements with their position while building the heap (in a similar way to foldMapWithIndex). As a result, there's a lot of code duplication.

**Comment by treeowl** [comment]
What happens if you actually *use* foldMapWithIndex?

On Jan 11, 2018 3:46 PM, "Donnacha Oisín Kidney" <notifications@github.com>
wrote:

Since the unstable sort is significantly quicker than the stable variant, I
think you can write a stable sort that uses it internally, and the overhead
of "stabilising" it should be small enough that it's still faster than
Data.List.sortBy.

I've written an implementation
<https://github.com/oisdk/containers/tree/faster-stable-sort> which uses a
pairing heap which tags each element with its original position:

data PQS e = PQS {-# UNPACK #-} !Int e (PQSL e)data PQSL e = Nl | {-#
UNPACK #-} !(PQS e) :&& PQSL e

It's quicker than the old implementation (which used Data.List.sort) in the
general case (all times are on sequences of length 50000):

Times (ms)            min    est    max  std dev   r²
to/from list:        64.23  64.50  64.81  0.432  1.000
1/11/18 stable heap: 38.87  39.40  40.09  0.457  0.999

Slightly slower in the case of already sorted lists:

Times (ms)            min    est    max  std dev   r²
to/from list:        6.806  6.861  6.901  0.234  1.000
1/11/18 stable heap: 8.211  8.268  8.328  0.111  1.000

And quicker in the case of lists sorted in reverse:

Times (ms)            min    est    max  std dev   r²
to/from list:        26.79  28.34  30.55  1.219  0.988
1/11/18 stable heap: 9.405  10.13  10.91  0.670  0.977

Interestingly, the stable sort is now competitive with the unstable:

Times (ms)            min    est    max  std dev   r²
unstable:            34.71  35.10  35.38  0.845  1.000
stable:              38.84  39.22  39.59  0.564  0.999

And even beats it in the case of already-sorted lists:

Times (ms)            min    est    max  std dev   r²
unstable:            8.457  8.499  8.536  0.069  1.000
stable:              8.160  8.230  8.334  0.158  0.999

I couldn't think of an way to actually use unstableSort: the new stable
sort tags elements with their position while building the heap (in a
similar way to foldMapWithIndex). As a result, there's a lot of code
duplication.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/haskell/containers/issues/491>, or mute the thread

**Comment by oisdk** [comment]
You can't use it in its current form: the pairing heap doesn't form a monoid without an ord constraint. Also, it doesn't have a `mempty` (an optimisation which slightly speeds up merging). I'll see what happens when foldMapWithIndex takes an explicit dictionary, and if I can put a `mempty` on the pairing heap somewhere

**Comment by treeowl** [comment]
It was just a thought. If writing directly is better, go ahead and do that.

**Comment by oisdk** [comment]
Yeah, it looks like you lose the speedup if you use foldMapWithIndex.

**Comment by bgamari** [comment]
It looks like this can be closed.

**Comment by treeowl** [comment]
Closed in #492

**Comment by treeowl** [comment]
Thanks.
